source/NRF52USB.cpp

```diff
@@ -0,0 +1,503 @@
+/*
+The MIT License (MIT)
+
+Copyright (c) 2017 Lancaster University.
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+*/
+#include "CodalUSB.h"
+#include "Timer.h"
+#include "Event.h"
+
+#if CONFIG_ENABLED(DEVICE_USB)
+#include "CodalDmesg.h"
+#include "nrfx_usbd.h"
+#include "nrfx_power.h"
+#include "nrfx_clock.h"
+#include "nrfx_usbd_errata.h"
+
+using namespace codal;
+
+#define LOG(...) ((void)0)
+// #define LOG DMESG
+#define DBG(...) ((void)0)
+// #define DBG DMESG
+
+#define NUM_IN_EP NRF_USBD_EPIN_CNT
+#define NUM_OUT_EP NRF_USBD_EPOUT_CNT
+
+#define CHK(call)                                                                                  \
+    do                                                                                             \
+    {                                                                                              \
+        int status = call;                                                                        \
+        usb_assert(status == NRFX_SUCCESS);                                                              \
+    } while (0)
+
+#if 0
+static UsbEndpointIn *findInEp(int ep)
+{
+    auto cusb = CodalUSB::usbInstance;
+    ep &= 0x7f;
+    if (ep == 0)
+        return cusb->ctrlIn;
+    for (auto iface = cusb->interfaces; iface; iface = iface->next)
+    {
+        if (iface->in && iface->in->ep == ep)
+            return iface->in;
+    }
+    return NULL;
+}
+
+static UsbEndpointOut *findOutEp(int ep)
+{
+    auto cusb = CodalUSB::usbInstance;
+    if (ep == 0)
+        return cusb->ctrlOut;
+    for (auto iface = cusb->interfaces; iface; iface = iface->next)
+    {
+        if (iface->out && iface->out->ep == ep)
+            return iface->out;
+    }
+    return NULL;
+}
+#endif
+
+#ifdef __cplusplus
```

Comment:
it's a .cpp file, you don't need this ifdef (of course still need the extern "C" though I would just put it in front of the IRQ handlers)

Reply:
Thank you! this was a habit I picked up from working with the armcc compiler back when we were using it as part of micro:bit. (It compiles both C and C++ files in the same compile step)